cranelift/native/src/riscv.rs

```diff
@@ -0,0 +1,126 @@
+    }
+
+    if (v & HWCAP_RISCV_EXT_F) != 0 {
+        isa_builder.enable("has_f").unwrap();
+
+        // TODO: There doesn't seem to be a bit associated with this extension
+        // rust enables it with the `f` extension:
+        // https://github.com/rust-lang/stdarch/blob/790411f93c4b5eada3c23abb4c9a063fb0b24d99/crates/std_detect/src/detect/os/linux/riscv.rs#L43
+        isa_builder.enable("has_zicsr").unwrap();
+    }
+
+    if (v & HWCAP_RISCV_EXT_M) != 0 {
+        isa_builder.enable("has_m").unwrap();
+    }
+
+    if (v & HWCAP_RISCV_EXT_V) != 0 {
+        isa_builder.enable("has_v").unwrap();
+    }
+
+    // In general extensions that are longer than one letter
+    // won't have a bit associated with them. The Linux kernel
+    // is currently working on a new way to query the extensions.
+    Ok(())
+}
+
+/// Read the /proc/cpuinfo file and detect the extensions.
+///
+/// We are looking for the isa line string, which contains the extensions.
+/// The format for this string is specifid in the linux user space ABI for RISC-V:
+/// https://github.com/torvalds/linux/blob/09a9639e56c01c7a00d6c0ca63f4c7c41abe075d/Documentation/riscv/uabi.rst
+///
+/// The format is fairly similar to the one specified in the RISC-V ISA manual, but
+/// all lower case.
+///
+/// An example ISA string is: rv64imafdcvh_zawrs_zba_zbb_zicbom_zicboz_zicsr_zifencei_zihintpause
+pub fn cpuinfo_detect(isa_builder: &mut dyn Configurable) -> Result<(), &'static str> {
+    let file = File::open("/proc/cpuinfo").map_err(|_| "failed to open /proc/cpuinfo")?;
+    let mut reader = BufReader::new(file);
+    let isa_line = reader
+        .by_ref()
+        .lines()
+        .find(|line| line.as_ref().map_or(false, |l| l.starts_with("isa")))
+        .ok_or("no isa line found in /proc/cpuinfo")?
+        .map_err(|_| "failed to read /proc/cpuinfo")?;
+
+    let isa_string = isa_line
+        .split(':')
+        .nth(1)
+        .ok_or("invalid isa string found in /proc/cpuinfo")?
+        .trim();
+
+    for ext in isa_string_extensions(isa_string) {
+        // Try enabling all the extensions that are parsed.
+        // Cranelift won't recognize all of them, but that's okay we just ignore them.
+        // Extensions flags in the RISC-V backend have the format of `has_x` for the `x` extension.
+        let _ = isa_builder.enable(&format!("has_{ext}"));
+    }
+    Ok(())
+}
+
+/// Parses an ISA string and returns an iterator over the extensions.
+fn isa_string_extensions(isa: &str) -> impl Iterator<Item = &str> {
+    isa.split('_').flat_map(|entry| {
+        // The first entry has the form `rv64imafdcvh`, we need to skip the first 4 characters.
+        // Each of the letters after the cpu architecture is an extension, so return them
+        // individually.
+        if entry.starts_with("rv") {
+            entry[4..].split("").filter(|e| !e.is_empty()).collect()
+        } else {
+            vec![entry]
+        }
+    })
+}
```

Comment:
Returning an iterator can make sense if it lets you avoid heap allocations, but in this case it requires allocating a `Vec` for each group. So I lean toward just accumulating one `Vec` for all the results and returning that. This isn't performance-critical, anyway.

I think I'd prefer to verify that the `rv64` part is the first one, which we can't cleanly do in an iterator chain, but can do by calling `Iterator::next` directly.

Finally, I fiddled around with different ways to get an iterator over single-character string slices. I like `s.matches(|_| true)` better than having to filter out empty strings from `str::split`.

```suggestion
 fn isa_string_extensions(isa: &str) -> Vec<&str> {
    let mut parts = isa.split('_');
    let mut extensions = Vec::new();
    // The first entry has the form `rv64imafdcvh`, we need to skip the architecture ("rv64").
    // Each of the letters after the cpu architecture is an extension, so return them
    // individually.
    if let Some(letters) = parts.next().unwrap().strip_prefix("rv64") {
        extensions.extend(letters.matches(|_| true));
        extensions.extend(parts);
    }
    extensions
}
```

Reply:
Yeah, I'm ok with that version as well. I tried to get a version that would just neatly return an iterator without allocating but couldn't get it. I left it in in the hopes that someone in the future would just see that and improve it.